### PR TITLE
Adding kubelet restart test for vsphere

### PR DESCRIPTION
**Pull request by divyenpatel**
**What this PR does / why we need it**:
Adding disruptive kubelet restart test for vsphere cloud provider.

Test Step.
This test verifies that a volume mounted to a pod remains mounted after a kubelet restarts.

1. Create volume (vmdk)
2. Create PV with volume path for the vmdk.
3. Create PVC to bind with PV.
4. Create a pod using the PVC.
5. Verify disk is attached to the node.
6. Write to the volume
7. Restart kubelet on the node on which pod is provisioned.
8. Verify that written file is accessible after kubelet restart.
9. Delete pod.
10. Delete PV, PVC
11. Wait for VMDK to detach from node.
12. Delete VMDK

**Special notes for your reviewer**:
func testKubeletRestartsAndRestoresMount is located in test/e2e/persistent_volumes-disruptive.go

**Release note**:

```release-note
None
```

cc: @kerneltime @abrarshivani @BaluDontu @tusharnt @pdhamdhere

### CODE REVIEW

**Comment by divyenpatel** [comment]
@jeffvance can you also review this

**Comment by jeffvance** [comment]
@divyenpatel yes, later this morning.

**Comment by jeffvance** [comment]
@k8s-bot test this

File: test/e2e/persistent_volumes-vsphere.go [hunk 1]
```diff
@@ -166,4 +166,14 @@ var _ = framework.KubeDescribe("PersistentVolumes:vsphere", func() {
 		By("Deleting the pod")
 		framework.DeletePodWithWait(f, c, clientPod)
 	})
+	/*
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-vsphere.go]: L33 `var _ = framework.KubeDescribe("PersistentVolumes:vsphere", func() {...`
should have "[Volume]" in the describe string, consistent with other persistent volumes tests.

Comment (jeffvance) [inline test/e2e/persistent_volumes-vsphere.go]: L101:
```
clientPod, err = c.CoreV1().Pods(ns).Get(clientPod.Name, metav1.GetOptions{})
if !apierrs.IsNotFound(err) { ...
```
What happens if err != nil but also != IsNotFound? I'd prefer to see:
```
clientPod, err = c.CoreV1().Pods(ns).Get(clientPod.Name, metav1.GetOptions{})
if err == nil {
     framework.DeletePodWithWait(f, c, clientPod)
} else {
     Expect(apierrs.IsNotFound(err)).To(BeTrue(), fmt.Sprintf("unexpected api error re-getting pod %q: %v", clientPod.Name, err))
}
```

File: test/e2e/persistent_volumes-vsphere.go [hunk 2]
```diff
@@ -166,4 +166,14 @@ var _ = framework.KubeDescribe("PersistentVolumes:vsphere", func() {
...
+		This test verifies that a volume mounted to a pod remains mounted after a kubelet restarts.
+		Steps:
+		1. Write to the volume
+		2. Restart kubelet
+		3. Verify that written file is accessible after kubelet restart
+	*/
+	It("should test that a file written to the mount before kubelet restart is stat-able after restart", func() {
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-vsphere.go]: I think "[Disruptive]" should be in this `It` text.

**Comment by jeffvance** [comment]
@k8s-bot non-cri node e2e test this

File: test/e2e/persistent_volumes-vsphere.go (hunk 1 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes-vsphere.go]: L101-103: 
```
clientPod, err = c.CoreV1().Pods(ns).Get(clientPod.Name, metav1.GetOptions{})
if !apierrs.IsNotFound(err) {
     framework.DeletePodWithWait(f, c, clientPod)
}
```
This handles unexpected errors better:
```
clientPod, err = c.CoreV1().Pods(ns).Get(clientPod.Name, metav1.GetOptions{})
if err == nil {
      framework.DeletePodWithWait(f, c, clientPod)
} else {
      Expect(apierrs.IsNotFound(err)).To(BeTrue(), fmt.Sprintf("unexpected api error re-getting pod %q: %v", clientPod.Name, err))
}
```

File: test/e2e/persistent_volumes-vsphere.go (hunk 2 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes-vsphere.go]: same comment ^^

**Comment by jeffvance** [comment]
@k8s-bot cvm gce e2e test this

File: test/e2e/persistent_volumes-vsphere.go (hunk 1 above)
Comment (divyenpatel) [inline test/e2e/persistent_volumes-vsphere.go]: Fixed. Handled err != nil case.

File: test/e2e/persistent_volumes-vsphere.go (hunk 2 above)
Comment (divyenpatel) [inline test/e2e/persistent_volumes-vsphere.go]: Added `[Disruptive]` tag to the `It` spec

**Comment by grodrigues3** [comment]
@jeffvance the bot sees your approval, it just needs someone from the test/OWNERS files.

**Comment by jeffvance** [comment]
@k8s-bot test this

**Comment by grodrigues3** [comment]
@jeffvance if you've reviewed and are happy with the code you (only you since you're the assignee) can type /lgtm

**Comment by divyenpatel** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this

**Comment by divyenpatel** [comment]
@jeffvance can you take a look?

**Comment by jeffvance** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this

**Comment by divyenpatel** [comment]
@k8s-bot unit test this

**Comment by jeffvance** [comment]
@k8s-bot unit test this

**Comment by jeffvance** [comment]
@k8s-bot cvm gce e2e test this

**Comment by divyenpatel** [comment]
closing this PR. added tests to PR https://github.com/kubernetes/kubernetes/pull/43143

**Pull request closed without merging**